File: jaxtari/jaxtari/diff/soft_select.py
"""Soft-selection primitive — softmax-weighted mixture over choices.

Paper reference: the second of the three relaxations in "Hard and Soft
Execution" — opcode dispatch as a convex combination over the handler
outputs, select(l, V; T) = w^T V with w = softmax(l / T) (Eq. "select";
supplementary "Setup and Notation", second primitive). The same softmax
select governs *any* discrete choice, with T setting how widely the
soft pick spreads (the "capture range" of Fig. "Soft-select temperature
T in pixel space"). The executed step uses the *saturated* (hard) form
of this select on the forward path, so dispatch is forward-exact
(Theorem 1); the relaxed form here is what the fully relaxed variant of
Theorem 2 ("Temperature-limit bound") uses. The temperature-limit proof
bounds the non-winning softmax mass by (K-1) exp(-Delta_min / T), so the
mixture collapses to the hard pick as T -> 0.

Mirrors the hard opcode dispatch of xitari M6502Low::execute (the
`switch(peekWithPC())` over the decoded opcode byte): here that switch
is expressed as a softmax-weighted mixture so gradients can flow into
the non-selected handlers.

PORTING_PLAN.md §6.2 calls this out for opcode dispatch ("a 256-way
softmax over the decoded opcode logits, with each branch's effect
weighted by its activation"). It's also the natural building block for
*any* if/else relaxation where the discrete pick can be expressed as
"which row of `values` should this be?"

Saturation: with very-large logits (one entry ≫ others) the softmax
collapses to one-hot and the result is bit-exact equal to the
hard-pick value. With moderate logits, gradient flows back through the
mix into the alternative branches.

A temperature argument lets the caller dial between "almost hard"
(small T) and "softer" (larger T). T → 0 forces argmax; T → ∞ forces
uniform.
"""

from __future__ import annotations

import jax
import jax.numpy as jnp


def soft_select(
    logits: jnp.ndarray,
    values: jnp.ndarray,
    temperature: float = 1.0,
) -> jnp.ndarray:
    """Compute `softmax(logits / T) · values`.

    `logits` and `values` must have the same leading dimension. `values`
    may have additional trailing dimensions; the result keeps them.
    """
    weights = jax.nn.softmax(logits / temperature)
    return jnp.tensordot(weights, values, axes=1)

File: jutari/src/diff/SoftSelect.jl
# Soft-selection primitive — softmax-weighted mixture over choices.
#
# Used for soft opcode dispatch and soft if/else relaxation.
# Saturation: with very-large logits the softmax collapses to one-hot
# and the result is bit-exact-equal to the hard-pick value. Lower
# temperature → harder pick; higher temperature → smoother mixture.
#
# Paper reference: the second relaxation in "Hard and Soft Execution" —
# opcode dispatch as a convex combination over handler outputs,
# select(l, V; T) = w' V with w = softmax(l / T) (Eq. "select";
# supplementary second primitive). The executed step uses the saturated
# (hard) form, so dispatch is forward-exact (Theorem 1); the relaxed
# form here is the one analysed in Theorem 2 ("Temperature-limit bound"),
# whose proof bounds the non-winning softmax mass by (K-1)exp(-Delta_min
# / T). Mirrors the hard opcode switch of xitari M6502Low::execute.

@inline function _softmax(logits::AbstractVector{<:Real})
    m = maximum(logits)
    e = exp.(logits .- m)
    return e ./ sum(e)
end

"""
    soft_select(logits, values; temperature=1.0)

Compute `softmax(logits / T) · values`. `logits` and `values` must
have matching leading dimension. If `values` is 2-D, the result is a
vector whose length equals `size(values, 2)`.
"""
function soft_select(logits::AbstractVector{<:Real},
                     values::AbstractVector{<:Real};
                     temperature::Real = 1.0)
    weights = _softmax(Float32.(logits) ./ Float32(temperature))
    return _dot(weights, Float32.(values))
end

function soft_select(logits::AbstractVector{<:Real},
                     values::AbstractMatrix{<:Real};
                     temperature::Real = 1.0)
    weights = _softmax(Float32.(logits) ./ Float32(temperature))
    return Float32.(values)' * weights
end
